### PR TITLE
Add "flipX / flipY" property to animations

### PR DESCRIPTION
**Pull request by larsiusprime**
...and also make it interact properly with FlxSprite's own flipX / flipY.

Rationale: for games with big sprite sheets, it can save a lot of memory to re-use frames in animations by vertically flipping them. So if I can re-use a walk cycle and just tell it to flip the frames right in the animation call, that's extremely convenient. Much easier to do than messing around with individual frames in the atlas.

Before this pull, we have had the ability to apply flipX/flipY to individual flxFrames (such as those loaded from an atlas) -- so this means the user can now set flipX/flipY in three places:

1) On the sprite itself
2) On an individual frame (e.g. tagging the metadata as part of atlas loading)
3) On an individual animation

The effect should cascade properly. So you can have a particular frame in an animation that has flipX=true, used in an animation that has flipX=true, and then the sprite itself has flipX=true, which will resolve to true because (true XOR true XOR true) == true.

### CODE REVIEW

**Comment by larsiusprime** [comment]
I tried that first, but it's only a numeric operator, not a boolean. There's no boolean xor.

**Comment by Gama11** [comment]
You could also just use `a != b`. ;)

File: flixel/FlxSprite.hx [hunk 1]
```diff
@@ -934,15 +934,16 @@ class FlxSprite extends FlxObject
...
+			var doFlipY:Bool = checkFlipY();
+			
 			if (!doFlipX && !doFlipY && _frame.type == FlxFrameType.REGULAR)
 			{
 				framePixels = _frame.paint(framePixels, _flashPointZero, false, true);
 			}
 			else
 			{
-				framePixels = _frame.paintRotatedAndFlipped(framePixels, _flashPointZero, FlxFrameAngle.ANGLE_0, flipX, flipY, false, true);
+				framePixels = _frame.paintRotatedAndFlipped(framePixels, _flashPointZero, FlxFrameAngle.ANGLE_0, checkFlipX(), checkFlipY(), false, true);
```
Comment (Gama11) [inline flixel/FlxSprite.hx]: Why call the check functions again here when the results are locally cached a few lines earlier in `doFlipX` / `Y`?

File: flixel/FlxSprite.hx [hunk 2]
```diff
@@ -1357,6 +1358,30 @@ class FlxSprite extends FlxObject
 	{
 		return antialiasing = value;
 	}
+	
+	private inline function checkFlipX():Bool
+	{
+		var doFlipX:Bool = FlxMath.boolXOR(flipX, _frame.flipX);
```
Comment (Gama11) [inline flixel/FlxSprite.hx]: There's not really much of a reason for this line to be always executed - it's just the else case of `if (animation.curAnim != null)`.

File: flixel/animation/FlxAnimation.hx [hunk 3]
```diff
@@ -49,6 +49,16 @@ class FlxAnimation extends FlxBaseAnimation
 	public var reversed(default, null):Bool = false;
 	
 	/**
+	 * Whether or not the frames of this animation are horizontally flipped
+	 */
+	public var flipX(default, null):Bool = false;
```
Comment (Gama11) [inline flixel/animation/FlxAnimation.hx]: Considering these are checked during each render, is there a reason to make them read-only? Setting them after creation should work.

**Comment by larsiusprime** [comment]
@gama11 how bout now?

**Comment by larsiusprime** [comment]
Ah, you're right. Well in any case the logic as it stands now is what I actually intend.

**Comment by Gama11** [comment]
Ah. Well, in that case, I'd prefer the original approach, since doing it this way introduces some redundancy. :P

**Comment by larsiusprime** [comment]
To be clear, this is what you mean?

```Haxe
	private inline function checkFlipX():Bool
	{
		var doFlipX = (flipX != _frame.flipX);
		if (animation.curAnim != null)
		{
			return (doFlipX != animation.curAnim.flipX);
		}
		return doFlipX;
	}
```

**Comment by Gama11** [comment]
Yes.

**Comment by larsiusprime** [comment]
Cool. Incoming.

**Comment by MSGhero** [comment]
Oh, right. Yeah this looks a bit nicer.

**Comment by larsiusprime** [comment]
Is this good to go then @Gama11 and @MSGhero or does it need more changes?

**Comment by MSGhero** [comment]
Why are flipX and Y parameters added only to a couple of the flxanimcontroller functions?

**Comment by larsiusprime** [comment]
The use case I have in mind is that you set those when you create a new animation, so I put it in the functions that create new FlxAnimations. Are there some others you think it should be added to? If so, which ones?

**Comment by MSGhero** [comment]
:+1: I guess it doesn't make sense adding it to `append()`.

**Comment by larsiusprime** [comment]
Ready for final review @MSGhero @Gama11

**Comment by MSGhero** [comment]
I can't test it right now, but it looks fine to me.

**Comment by larsiusprime** [comment]
Bump for final review?
@Beeblerox @gamedevsam @Gama11

**Pull request merged**